### ISSUE TITLE
codeql query compile error

### ISSUE THREAD

**Issue by Cirengg** [github/codeql#12361]
I am running WHQL for Windows 11 22H2 which requires the output of running CodeQL on my source code.
So I followed the instructions given [here](https://github.com/microsoft/Windows-Driver-Developer-Supplemental-Tools) to setup CodeQL and bumped into these problems:

**1. codeql/suite-helpers@0.4.3 does not exist**
I did a `codeql pack download codeql/cpp-queries` to download the required `cpp-queries` QL pack. Then navigated to the folder containing qlpack.yml and ran `codeql pack install`. 

Got this output:
`A fatal error occurred: Package version failure. No compatible version found: codeql/suite-helpers@>=0.4.3 <=0.4.3.`

Apparently, the latest version on the registry is 0.4.1. Is there something broken?
To get around this problem, I have to download `codeql/cpp-queries@0.5.1` instead.

**2. compilation error**
With cpp-queries@0.5.1, the database analyze command proceeded to compile the query and came back with lots of errors which I don't know how to resolve. Please help!

Here is what I did:
`codeql database analyze C:\codeql-home\Databases\kmdf --format=sarifv2.1.0 --output=C:\codeql-home\AnalysisResults\KMDFResults.sarif C:\codeql-home\Windows-Driver-Developer-Supplemental-Tools\codeql\windows-drivers\queries\suites\windows_driver_mustfix.qls`

Here is what I got back:
```
Running queries.
Compiling query plan for C:\Users\dev\.codeql\packages\codeql\cpp-queries\0.5.1\Likely Bugs\Arithmetic\BadAdditionOverflowCheck.ql.
ERROR: unexpected input ';' expecting one of: 'instanceof', '{' (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:209,66-67)
ERROR: unexpected input '<' expecting one of: '=', '{' (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:222,38-39)
ERROR: unexpected input 'Target' expecting one of: Lowerid (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:222,64-70)
ERROR: unexpected input '>' expecting one of: Lowerid (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:222,70-71)
ERROR: unexpected input '<' expecting one of: '=', '{' (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:258,14-15)
ERROR: unexpected input 'P' expecting one of: Lowerid (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:258,21-22)
ERROR: unexpected input '>' expecting one of: Lowerid (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:258,22-23)
ERROR: unexpected input '}' expecting one of: <EOF>, 'boolean', 'class', 'newtype', 'date', 'float', 'from', 'import', 'int', 'module', 'predicate', 'select', 'string', 'where', Lowerid, Upperid, Atlowerid, '/**' (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:315,3-4)
ERROR: Could not resolve module P (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:259,20-21)
ERROR: Could not resolve type Target (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:229,26-32)
ERROR: Could not resolve type Target (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:236,20-26)
ERROR: Could not resolve type Target (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:242,21-27)
ERROR: Could not resolve type Target (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:250,31-37)
ERROR: Could not resolve type Target (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:265,7-13)
ERROR: Could not resolve type Target (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:222,64-70)
ERROR: Could not resolve type Param (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:258,15-20)
ERROR: Could not resolve type P (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:258,21-22)
ERROR: Could not resolve predicate getReallocPtrArg/1 (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:271,18-34)
ERROR: Could not resolve predicate getSizeArg/1 (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:272,28-38)
ERROR: Could not resolve predicate getSizeArg/1 (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:281,61-71)
ERROR: Could not resolve predicate getSizeMult/1 (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:282,21-32)
ERROR: Could not resolve predicate getSizeMult/1 (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:294,36-47)
ERROR: Could not resolve predicate getSizeMult/1 (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:297,20-31)
ERROR: Could not resolve predicate getSizeArg/1 (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:298,47-57)
ERROR: Could not resolve predicate getReallocPtrArg/1 (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:305,61-77)
ERROR: Could not resolve predicate requiresDealloc/1 (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:313,41-56)
ERROR: Declaration of predicate () conflicts with existing declarations: predicate () from Allocation.qll:222, predicate () from Allocation.qll:258, predicate () from Allocation.qll:258 (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:222,64-70)
ERROR: Declaration of predicate () conflicts with existing declarations: predicate () from Allocation.qll:222, predicate () from Allocation.qll:258, predicate () from Allocation.qll:258 (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:222,70-73)
ERROR: Declaration of predicate () conflicts with existing declarations: predicate () from Allocation.qll:222, predicate () from Allocation.qll:222, predicate () from Allocation.qll:258 (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:258,21-22)
ERROR: Declaration of predicate () conflicts with existing declarations: predicate () from Allocation.qll:222, predicate () from Allocation.qll:222, predicate () from Allocation.qll:258 (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:258,22-25)
ERROR: unknown annotation signature (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:224,3-12)
ERROR: this predicate requires a body (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:229,9-25)
ERROR: this predicate requires a body (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:236,9-19)
ERROR: this predicate requires a body (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:242,9-20)
ERROR: this predicate requires a body (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:250,15-30)
ERROR: unknown annotation signature (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:209,9-18)
ERROR: this predicate requires a body (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:222,64-70)
ERROR: this predicate requires a body (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:222,70-73)
ERROR: this predicate requires a body (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:258,21-22)
ERROR: this predicate requires a body (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:258,22-25)
WARNING: Unused variable result (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:229,5-8)
WARNING: Unused variable target (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:229,33-39)
WARNING: Unused variable result (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:236,5-8)
WARNING: Unused variable target (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:236,27-33)
WARNING: Unused variable result (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:242,5-8)
WARNING: Unused variable target (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:242,28-34)
WARNING: Unused variable target (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:250,38-44)
WARNING: Unused variable result (C:\Users\dev\.codeql\packages\codeql\cpp-all\0.5.1\semmle\code\cpp\models\implementations\Allocation.qll:222,39-63)
```

**Comment by hmakholm** [comment]
The instructions you're using are not really internally consistent. They tell you to use version 2.4.x or 2.6.x, which are both fairly old, but then instruct you to download the _latest_ version of the query packs with `pack install`. The QL language has evolved since version 2.6.x, and the current queries use some language features that those old CLI releases don't understand -- thus all the compiler errors.

The smoothest advice I can get for getting queries to run would be to use a newer version of the CLI. Probably the 2.11.5 that was used to write the instructions will still work -- but since you'll be using the _latest_ queries, it will really be more robust to grab the latest CodeQL CLI too.

If the group at MS that does driver certification really wants you to use those old CodeQL versions, you'll have to get compatible query sources for them. You should pester the MS people to make sure their instructions actually work -- but in the mean time, instead of `codeql pack download` you can also clone https://github.com/github/codeql and check out the `codeql-cli/v2.6.3` or `codeql-cli/v2.4.6` tag. Then use the `--additional-packs` flag in the `codeql database analyze` to point to that checkout.

---

Using an old version of CodeQL is most likely also the reason for your trouble with the `suite-helpers` pack. That suite-helpers pack is included within the `cpp-queries` pack you download as a bundled library dependency -- but old versions of the CLI had a bug where they wouldn't _use_ the bundled dependency in the particular way it is being used, and instead wanted to download it separately. Once that bug was fixed (and after giving a bit of time for everyone to update), we stopped publishing new suite-helpers packs separately to the registry ...

**Comment by Cirengg** [comment]
Hi @hmakholm 

Indeed, the version of CLI is the root cause and the instruction given on the page is sort of obsolete. Should have used the main branch and ignored the other WHCP_XXXX branches altogether. Problem was resolved with 2.11.5 CLI.
Thanks! I will close the issue.
